src/Test/winswTests/ServiceDescriptorTests.cs

```diff
@@ -441,5 +444,20 @@ public void DelayedStart_RoundTrip(bool enabled)
             var sd = bldr.ToServiceDescriptor();
             Assert.That(sd.DelayedAutoStart, Is.EqualTo(enabled));
         }
+
+        [Test]
+        public void ValidateAndLoadXmlSchemaTest()
+        {
+            const string seedXml = "<service>" +
+  "<id>myapp</id> " +
+  "<name>MyApp Service (powered by WinSW)</name>" +
+  "<description>This service is a service created from a minimal configuration</description>" +
+"</service>";
+
+            var serviceDescriptor = ServiceDescriptor.FromXML(seedXml);
+
+            var reader = XmlReader.Create(new StringReader(seedXml));
+            serviceDescriptor.ValidateAndLoadXmlSchema(reader);

```

Comment:
Now let's focus on the test. The validator's job is to detect invalid XML pieces and throw exceptions for them. In order to verify it works there needs to be a test to ensure that it does throw an exception.

You can write an assertion like `Assert.That(() => DoSomething(), Throws.TypeOf<XmlException>());`.

Reply:
Thank you for the description. You are so helping. :)
Here this part is not clear to me. I'm testing here, whether my function which is tested is thrown any **XmlException**. If this throws any Exception then my test will be passed, otherwise test will be failed. So I think I need to feed XML string as the test case, which fails the validation, then exception will be thrown and Test will pass. Am I correct?

Reply:
Right. You feed an invalid XML config string (valid XML but invalid WinSW config) and expect the validator to throw.

Reply:
This is the XSD, I have provided.

<xs:element name="service">
    <xs:complexType>
      <xs:sequence>
        <xs:element name="id" type="xs:string"/>
        <xs:element name="name" type="xs:string"/>
        <xs:element name="description" type="xs:string"/>
        <xs:element name="executable" type="xs:string"/>
      </xs:sequence>
    </xs:complexType>
  </xs:element>

But still validation don't catch error in below XML.

<id>
  <service>myapp</service>
  <name>MyApp Service (powered by WinSW)</name>
  <description>This service is a service created from a minimal configuration</description>
</id>

So how to what kind of a XML will fail in this condition. I think obviously this configuration xml is invalid. But XML validator not catching it.

Reply:
Hmm, it's definitely odd, right? Just a heads-up, the `XmlReaderSettings` you create won't automatically apply to newly constructed `XmlReader`s.

Reply:
Ooh.. I have missed it.  I change method like this.

```
public void ValidateAndLoadXmlSchema(XmlReader reader)
        {
            XmlReaderSettings settings =  new XmlReaderSettings();

            settings.ValidationType = ValidationType.Schema;
            settings.ValidationEventHandler += new ValidationEventHandler(XmlValidationEventHandler);

            Assembly a = Assembly.GetExecutingAssembly();

            using (Stream schemaStream = a.GetManifestResourceStream("winsw.XMLSchema.xsd"))
            {
                using (XmlReader schemaReader = XmlReader.Create(schemaStream))
                {
                    settings.Schemas.Add(null, schemaReader);
                }
            }

            var new_reader = XmlReader.Create(reader, settings);

            dom.Load(new_reader);
        }
```
But still it fails to catch the validation error.

Reply:
`dom` has been loaded in `ServiceDescriptor.FromXML` so it refuses to load again.

Reply:
I add another constructor to ServiceDescriptor and it works. But, since default ctor already there I create it with taking integer as a argument and I pass 0 into that in test and run XML validation method from ServiceDescriptor object created with that constructor. But I don't know whether it is a good practice. Shall I commit that ?